backend/amd64/cfg_selection.ml

```diff
@@ -224,241 +224,226 @@ let pseudoregs_for_operation op arg res =
-      Misc.fatal_error "is_store_out_of_range should not be called on x86-64"
-  end
-
-let fundecl ~future_funcnames f =
-  Cfg.reset_instr_id ();
-  (new selector)#emit_fundecl ~future_funcnames f
+    raise Use_default_exn
+
+let is_immediate (op : Simple_operation.integer_operation) n :
+    Cfg_selectgen_target_intf.is_immediate_result =
+  match op with
+  | Iadd | Isub | Imul | Iand | Ior | Ixor | Icomp _ ->
+    Is_immediate (is_immediate_int n)
+  | _ -> Use_default
+
+let is_immediate_test _cmp n : Cfg_selectgen_target_intf.is_immediate_result =
+  Is_immediate (is_immediate_int n)
+
+let is_simple_expr (expr : Cmm.expression) :
+    Cfg_selectgen_target_intf.is_simple_expr_result =
+  match expr with
+  | Cop (Cextcall { func = fn }, args, _) when List.mem fn inline_ops ->
+    (* inlined ops are simple if their arguments are *)
+    Simple_if_all_expressions_are args
+  | _ -> Use_default
+
+let effects_of (expr : Cmm.expression) :
+    Cfg_selectgen_target_intf.effects_of_result =
+  match expr with
+  | Cop (Cextcall { func = fn }, args, _) when List.mem fn inline_ops ->
+    Effects_of_all_expressions args
+  | _ -> Use_default
+
+let select_addressing (_chunk : Cmm.memory_chunk) exp :
+    addressing_mode * Cmm.expression =
+  let a, d = select_addr exp in
+  (* PR#4625: displacement must be a signed 32-bit immediate *)
+  if not (is_immediate_int d)
+  then Iindexed 0, exp
+  else
+    match a with
+    | Asymbol s ->
+      let glob : Arch.sym_global =
+        match s.sym_global with Global -> Global | Local -> Local
+      in
+      Ibased (s.sym_name, glob, d), Ctuple []
+    | Alinear e -> Iindexed d, e
+    | Aadd (e1, e2) -> Iindexed2 d, Ctuple [e1; e2]
+    | Ascale (e, scale) -> Iscaled (scale, d), e
+    | Ascaledadd (e1, e2, scale) -> Iindexed2scaled (scale, d), Ctuple [e1; e2]
+
+let select_store ~is_assign addr (exp : Cmm.expression) :
+    Cfg_selectgen_target_intf.select_store_result =
+  match exp with
+  | Cconst_int (n, _dbg) when is_immediate_int n ->
+    Rewritten
+      (Specific (Istore_int (Nativeint.of_int n, addr, is_assign)), Ctuple [])
+  | Cconst_natint (n, _dbg) when is_immediate_natint n ->
+    Rewritten (Specific (Istore_int (n, addr, is_assign)), Ctuple [])
+  | Cconst_int _ | Cconst_vec128 _
+  | Cconst_natint (_, _)
+  | Cconst_float32 (_, _)
+  | Cconst_float (_, _)
+  | Cconst_symbol (_, _)
+  | Cvar _
+  | Clet (_, _, _)
+  | Cphantom_let (_, _, _)
+  | Ctuple _
+  | Cop (_, _, _)
+  | Csequence (_, _)
+  | Cifthenelse (_, _, _, _, _, _, _)
+  | Cswitch (_, _, _, _, _)
+  | Ccatch (_, _, _, _)
+  | Cexit (_, _, _)
+  | Ctrywith (_, _, _, _, _, _, _) ->
+    Use_default
+
+let is_store_out_of_range _chunk ~byte_offset:_ :
+    Cfg_selectgen_target_intf.is_store_out_of_range_result =
+  Within_range
+
+let insert_move_extcall_arg _exttype _src _dst :
+    Cfg_selectgen_target_intf.insert_move_extcall_arg_result =
+  Use_default
```

Comment:
(I prefer this non-raising version, but
before you changed it I was wondering
whether the interface should have
optional functions. The version of this
pull request is probably less noisy.)